File: frontend/src/api/expenses.ts
import * as FileSystem from "expo-file-system/legacy";

import { apiClient, getApiAuthHeaders, resolveReachableApiUrlCandidates } from "@/api/client";
import type {
  ExpenseEntryCreate,
  ExpenseEntryPage,
  ExpenseEntryRead,
  ExpenseItemCounts,
  ExpenseItemCreate,
  ExpenseItemRead,
  ExpenseItemRowsPage,
  ExpenseItemUpdate,
  ShopExpenseAllocationBulkCreate,
  ShopExpenseAllocationBulkRead,
  ShopExpenseAllocationUpdate,
  ShopExpenseItemRead,
  ShopExpenseItemRowsPage,
  ShopExpenseItemsOrderRead,
  ShopExpenseItemsOrderUpdate,
  UUID,
} from "@/types/api";

export type ExpenseItemImageUploadFile = {
  uri: string;
  name: string;
  type: string;
};

export type ExpenseCursorParams = {
  q?: string;
  active?: boolean | null;
  limit?: number;
  cursor_sort_order?: number | null;
  cursor_name?: string | null;
  cursor_id?: UUID | null;
};

export type ExpenseHistoryParams = {
  shop_id?: UUID | null;
  range_start_date?: string | null;
  range_end_date?: string | null;
  limit?: number;
  cursor_spent_at?: string | null;
  cursor_id?: UUID | null;
};

function rowParams(params?: ExpenseCursorParams) {
  return {
    q: params?.q || undefined,
    active: params?.active ?? undefined,
    limit: params?.limit ?? 50,
    cursor_sort_order: params?.cursor_sort_order ?? undefined,
    cursor_name: params?.cursor_name ?? undefined,
    cursor_id: params?.cursor_id ?? undefined,
  };
}

function historyParams(params?: ExpenseHistoryParams) {
  return {
    shop_id: params?.shop_id ?? undefined,
    range_start_date: params?.range_start_date ?? undefined,
    range_end_date: params?.range_end_date ?? undefined,
    limit: params?.limit ?? 50,
    cursor_spent_at: params?.cursor_spent_at ?? undefined,
    cursor_id: params?.cursor_id ?? undefined,
  };
}

function parseUploadResponseBody(body: string) {
  try {
    return body ? JSON.parse(body) : {};
  } catch {
    return {};
  }
}

function getUploadResponseMessage(body: unknown) {
  if (!body || typeof body !== "object") {
    return "";
  }
  const detail = (body as { detail?: unknown }).detail;
  if (typeof detail === "string") {
    return detail;
  }
  const message = (body as { message?: unknown }).message;
  return typeof message === "string" ? message : "";
}

async function assertUploadFileReady(file: ExpenseItemImageUploadFile) {
  const info = await FileSystem.getInfoAsync(file.uri);
  if (!info.exists || info.isDirectory) {
    throw new Error("Selected image file is no longer available. Pick the image again and save.");
  }
}

async function uploadExpenseItemImageFile(path: string, file: ExpenseItemImageUploadFile) {
  const uploadUrls = await resolveReachableApiUrlCandidates(path);
  if (uploadUrls.length === 0) {
    throw new Error("API base URL is not configured. Set EXPO_PUBLIC_API_BASE_URL and restart Expo.");
  }
  await assertUploadFileReady(file);

  let lastError: unknown = null;
  for (const uploadUrl of uploadUrls) {
    try {
      const response = await FileSystem.uploadAsync(uploadUrl, file.uri, {
        fieldName: "image",
        headers: {
          Accept: "application/json",
          ...getApiAuthHeaders(),
        },
        httpMethod: "PUT",
        mimeType: file.type,
        uploadType: FileSystem.FileSystemUploadType.MULTIPART,
      });
      const body = parseUploadResponseBody(response.body);
      if (response.status >= 200 && response.status < 300) {
        return body as ExpenseItemRead;
      }
      throw new Error(getUploadResponseMessage(body) || `Image upload failed with status ${response.status}.`);
    } catch (error) {
      lastError = error;
    }
  }

  if (lastError instanceof Error) {
    throw lastError;
  }
  throw new Error("Image upload failed before the backend responded.");
}

export async function fetchExpenseItemRows(
  params?: ExpenseCursorParams,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ExpenseItemRowsPage>("/api/v1/admin/expenses/items", {
    params: rowParams(params),
    signal: options.signal,
  });
  return data;
}

export async function fetchExpenseItemCounts(
  params?: Pick<ExpenseCursorParams, "q">,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ExpenseItemCounts>("/api/v1/admin/expenses/items/counts", {
    params: { q: params?.q || undefined },
    signal: options.signal,
  });
  return data;
}

export async function createExpenseItem(payload: ExpenseItemCreate) {
  const { data } = await apiClient.post<ExpenseItemRead>("/api/v1/admin/expenses/items", payload);
  return data;
}

export async function updateExpenseItem(itemId: UUID, payload: ExpenseItemUpdate) {
  const { data } = await apiClient.patch<ExpenseItemRead>(`/api/v1/admin/expenses/items/${itemId}`, payload);
  return data;
}

export async function deleteExpenseItem(itemId: UUID) {
  await apiClient.delete(`/api/v1/admin/expenses/items/${itemId}`);
}

export async function replaceExpenseItemImageFile(itemId: UUID, file: ExpenseItemImageUploadFile) {
  return uploadExpenseItemImageFile(`/api/v1/admin/expenses/items/${itemId}/image`, file);
}

export async function deleteExpenseItemImage(itemId: UUID) {
  const { data } = await apiClient.delete<ExpenseItemRead>(`/api/v1/admin/expenses/items/${itemId}/image`);
  return data;
}

export async function fetchShopExpenseItemRows(
  shopId: UUID,
  params?: ExpenseCursorParams,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ShopExpenseItemRowsPage>(
    `/api/v1/admin/shops/${shopId}/expense-items`,
    {
      params: rowParams(params),
      signal: options.signal,
    },
  );
  return data;
}

export async function fetchShopExpenseItemCounts(
  shopId: UUID,
  params?: Pick<ExpenseCursorParams, "q">,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ExpenseItemCounts>(
    `/api/v1/admin/shops/${shopId}/expense-items/counts`,
    {
      params: { q: params?.q || undefined },
      signal: options.signal,
    },
  );
  return data;
}

export async function fetchShopExpenseItemCandidateRows(
  shopId: UUID,
  params?: ExpenseCursorParams,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ExpenseItemRowsPage>(
    `/api/v1/admin/shops/${shopId}/expense-item-candidates`,
    {
      params: rowParams(params),
      signal: options.signal,
    },
  );
  return data;
}

export async function allocateShopExpenseItems(shopId: UUID, itemIds: UUID[]) {
  const payload: ShopExpenseAllocationBulkCreate = { expense_item_ids: itemIds };
  const { data } = await apiClient.post<ShopExpenseAllocationBulkRead>(
    `/api/v1/admin/shops/${shopId}/expense-items/allocations`,
    payload,
  );
  return data;
}

export async function allocateShopExpenseItem(shopId: UUID, itemId: UUID) {
  const { data } = await apiClient.post<ShopExpenseItemRead>(
    `/api/v1/admin/shops/${shopId}/expense-items/${itemId}/allocation`,
  );
  return data;
}

export async function updateShopExpenseAllocation(
  shopId: UUID,
  itemId: UUID,
  payload: ShopExpenseAllocationUpdate,
) {
  const { data } = await apiClient.patch<ShopExpenseItemRead>(
    `/api/v1/admin/shops/${shopId}/expense-items/${itemId}/allocation`,
    payload,
  );
  return data;
}

export async function deallocateShopExpenseItem(shopId: UUID, itemId: UUID) {
  const { data } = await apiClient.delete<ShopExpenseItemRead>(
    `/api/v1/admin/shops/${shopId}/expense-items/${itemId}/allocation`,
  );
  return data;
}

export async function updateShopExpenseItemsOrder(shopId: UUID, payload: ShopExpenseItemsOrderUpdate) {
  const { data } = await apiClient.put<ShopExpenseItemsOrderRead>(
    `/api/v1/admin/shops/${shopId}/expense-items/order`,
    payload,
  );
  return data;
}

export async function fetchAdminExpenseHistory(
  params?: ExpenseHistoryParams,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ExpenseEntryPage>("/api/v1/admin/expenses/history", {
    params: historyParams(params),
    signal: options.signal,
  });
  return data;
}

export async function fetchCurrentShopExpenseItems(
  params?: ExpenseCursorParams,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ShopExpenseItemRowsPage>("/api/v1/shop/expenses/items", {
    params: rowParams(params),
    signal: options.signal,
  });
  return data;
}

export async function createShopExpenseEntry(payload: ExpenseEntryCreate) {
  const { data } = await apiClient.post<ExpenseEntryRead>("/api/v1/shop/expenses/entries", payload);
  return data;
}

export async function fetchShopExpenseHistory(
  params?: Omit<ExpenseHistoryParams, "shop_id">,
  options: { signal?: AbortSignal } = {},
) {
  const { data } = await apiClient.get<ExpenseEntryPage>("/api/v1/shop/expenses/history", {
    params: historyParams(params),
    signal: options.signal,
  });
  return data;
}

File: backend/app/routers/admin/expenses.py
from app.routers.admin._common import *
from app.routers.admin._params import *

router = APIRouter()

@router.get(
    "/expenses/items",
    response_model=ExpenseItemRowsPage,
    response_model_exclude_unset=True,
    summary="List Expense Items",
)
async def get_expense_items(
    db: DBSession,
    q: ItemSearchParam = None,
    active: ItemActiveParam = None,
    limit: ItemsLimitParam = 50,
    cursor_sort_order: ItemCursorSortOrderParam = None,
    cursor_name: ItemCursorNameParam = None,
    cursor_id: ItemCursorIdParam = None,
) -> ExpenseItemRowsPage:
    return await list_expense_item_rows(
        db,
        q=q,
        active=active,
        limit=limit,
        cursor_sort_order=cursor_sort_order,
        cursor_name=cursor_name,
        cursor_id=cursor_id,
    )


@router.get(
    "/expenses/items/counts",
    response_model=ExpenseItemCounts,
    response_model_exclude_unset=True,
    summary="Count Expense Items",
)
async def get_expense_item_counts(
    db: DBSession,
    q: ItemSearchParam = None,
) -> ExpenseItemCounts:
    return await count_expense_items(db, q=q)


@router.post(
    "/expenses/items",
    response_model=ExpenseItemRead,
    response_model_exclude_unset=True,
    status_code=201,
    summary="Create Expense Item",
)
async def create_admin_expense_item(
    payload: ExpenseItemCreate,
    db: DBSession,
) -> ExpenseItemRead:
    return await create_expense_item(db, payload)


@router.get(
    "/expenses/items/{expense_item_id}",
    response_model=ExpenseItemRead,
    response_model_exclude_unset=True,
    summary="Get Expense Item",
)
async def get_admin_expense_item(expense_item_id: UUID, db: DBSession) -> ExpenseItemRead:
    return await get_expense_item(db, expense_item_id)


@router.patch(
    "/expenses/items/{expense_item_id}",
    response_model=ExpenseItemRead,
    response_model_exclude_unset=True,
    summary="Update Expense Item",
)
async def update_admin_expense_item(
    expense_item_id: UUID,
    payload: ExpenseItemUpdate,
    db: DBSession,
) -> ExpenseItemRead:
    return await update_expense_item(db, expense_item_id, payload)


@router.delete(
    "/expenses/items/{expense_item_id}",
    status_code=204,
    summary="Delete Expense Item",
)
async def delete_admin_expense_item(expense_item_id: UUID, db: DBSession) -> Response:
    await delete_expense_item(db, expense_item_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.put(
    "/expenses/items/{expense_item_id}/image",
    response_model=ExpenseItemRead,
    response_model_exclude_unset=True,
    summary="Upload Expense Item Image",
)
async def upload_admin_expense_item_image(
    expense_item_id: UUID,
    db: DBSession,
    image: ItemImageUploadRequired,
) -> ExpenseItemRead:
    return await upload_expense_item_image(db, expense_item_id, image)


@router.delete(
    "/expenses/items/{expense_item_id}/image",
    response_model=ExpenseItemRead,
    response_model_exclude_unset=True,
    summary="Delete Expense Item Image",
)
async def delete_admin_expense_item_image(expense_item_id: UUID, db: DBSession) -> ExpenseItemRead:
    return await remove_expense_item_image(db, expense_item_id)


@router.get(
    "/shops/{shop_id}/expense-items",
    response_model=ShopExpenseItemRowsPage,
    response_model_exclude_unset=True,
    summary="List Branch Expense Items",
)
async def get_shop_expense_items(
    shop: ShopDep,
    db: DBSession,
    q: ItemSearchParam = None,
    active: ItemActiveParam = None,
    limit: ItemsLimitParam = 50,
    cursor_sort_order: ItemCursorSortOrderParam = None,
    cursor_name: ItemCursorNameParam = None,
    cursor_id: ItemCursorIdParam = None,
) -> ShopExpenseItemRowsPage:
    return await list_shop_expense_item_rows(
        db,
        shop,
        q=q,
        active=active,
        limit=limit,
        cursor_sort_order=cursor_sort_order,
        cursor_name=cursor_name,
        cursor_id=cursor_id,
    )


@router.get(
    "/shops/{shop_id}/expense-items/counts",
    response_model=ExpenseItemCounts,
    response_model_exclude_unset=True,
    summary="Count Branch Expense Items",
)
async def get_shop_expense_item_counts(
    shop: ShopDep,
    db: DBSession,
    q: ItemSearchParam = None,
) -> ExpenseItemCounts:
    return await count_expense_items(db, shop_id=shop.id, q=q)


@router.get(
    "/shops/{shop_id}/expense-item-candidates",
    response_model=ExpenseItemRowsPage,
    response_model_exclude_unset=True,
    summary="List Branch Expense Item Candidates",
)
async def get_shop_expense_item_candidates(
    shop: ShopDep,
    db: DBSession,
    q: ItemSearchParam = None,
    limit: ItemsLimitParam = 50,
    cursor_sort_order: ItemCursorSortOrderParam = None,
    cursor_name: ItemCursorNameParam = None,
    cursor_id: ItemCursorIdParam = None,
) -> ExpenseItemRowsPage:
    return await list_shop_expense_candidate_rows(
        db,
        shop,
        q=q,
        limit=limit,
        cursor_sort_order=cursor_sort_order,
        cursor_name=cursor_name,
        cursor_id=cursor_id,
    )


@router.post(
    "/shops/{shop_id}/expense-items/allocations",
    response_model=ShopExpenseAllocationBulkRead,
    response_model_exclude_unset=True,
    status_code=201,
    summary="Allocate Branch Expense Items",
)
async def allocate_shop_expenses(
    payload: ShopExpenseAllocationBulkCreate,
    shop: ShopDep,
    db: DBSession,
) -> ShopExpenseAllocationBulkRead:
    return await allocate_shop_expense_items(db, shop, payload.expense_item_ids)


@router.post(
    "/shops/{shop_id}/expense-items/{expense_item_id}/allocation",
    response_model=ShopExpenseItemRead,
    response_model_exclude_unset=True,
    status_code=201,
    summary="Allocate Branch Expense Item",
)
async def allocate_shop_expense(
    expense_item_id: UUID,
    shop: ShopDep,
    db: DBSession,
) -> ShopExpenseItemRead:
    return await allocate_shop_expense_item(db, shop, expense_item_id)


@router.patch(
    "/shops/{shop_id}/expense-items/{expense_item_id}/allocation",
    response_model=ShopExpenseItemRead,
    response_model_exclude_unset=True,
    summary="Update Branch Expense Allocation",
)
async def update_shop_expense(
    expense_item_id: UUID,
    payload: ShopExpenseAllocationUpdate,
    shop: ShopDep,
    db: DBSession,
) -> ShopExpenseItemRead:
    return await update_shop_expense_allocation(db, shop, expense_item_id, payload)


@router.delete(
    "/shops/{shop_id}/expense-items/{expense_item_id}/allocation",
    response_model=ShopExpenseItemRead,
    response_model_exclude_unset=True,
    summary="Remove Branch Expense Allocation",
)
async def deallocate_shop_expense(
    expense_item_id: UUID,
    shop: ShopDep,
    db: DBSession,
) -> ShopExpenseItemRead:
    return await deallocate_shop_expense_item(db, shop, expense_item_id)


@router.put(
    "/shops/{shop_id}/expense-items/order",
    response_model=ShopExpenseItemsOrderRead,
    response_model_exclude_unset=True,
    summary="Update Branch Expense Item Order",
)
async def update_shop_expense_order(
    payload: ShopExpenseItemsOrderUpdate,
    shop: ShopDep,
    db: DBSession,
) -> ShopExpenseItemsOrderRead:
    return await update_shop_expense_items_order(db, shop, payload.expense_item_ids)


@router.patch(
    "/expenses/history/{entry_id}",
    response_model=ExpenseEntryRead,
    response_model_exclude_unset=True,
    summary="Update Expense History Entry",
)
async def update_admin_expense_entry(
    entry_id: UUID,
    payload: ExpenseEntryUpdate,
    db: DBSession,
) -> ExpenseEntryRead:
    return await update_expense_entry(db, entry_id, payload)


@router.get(
    "/expenses/history",
    response_model=ExpenseEntryPage,
    response_model_exclude_unset=True,
    summary="List Expense History",
)
async def get_expense_history(
    db: DBSession,
    shop_id: ShopIdParam = None,
    range_start_date: RangeStartDateParam = None,
    range_end_date: RangeEndDateParam = None,
    limit: ItemsLimitParam = 50,
    cursor_spent_at: CursorSpentAtParam = None,
    cursor_id: CursorIdParam = None,
) -> ExpenseEntryPage:
    return await list_expense_entries(
        db,
        shop_id=shop_id,
        range_start_date=range_start_date,
        range_end_date=range_end_date,
        limit=limit,
        cursor_spent_at=cursor_spent_at,
        cursor_id=cursor_id,
    )

